Instantiated class

softwx opened the pull request:
Converted Symspell class to be an instantiated class instead of a static class. Removed language feature/parameter (no longer needed since a separate SymSpell instance can be made for different languages). Now that it can be instantiated, it's easier to set up a benchmark suite, so also added a project for that. That should help evaluate/compare future optimisations. Added some more unit tests to help ensure that future enhancements don't introduce bugs.
This is a big change, so don't know if you're ready to pull it into master, but it's ready to review.

wolfgarbe commented:
Thanks for the great refactoring from static to instantiated class!

softwx commented:
Wolf,
I've worked out a design for a dictionary structure that allows using
string hashcodes to directly index into the itemList (getting rid of the
need for the dictionary collection), yet still allow for dynamically adding
words and growing the structure with a very low cost (fast and simple) size
doubling  + re-hashing when needed. I think it will be both faster and more
space efficient, but won't know how well and how much until I implement it.
The design would also allow some control of the speed vs. space preference.
It may be easier and cleaner to implement if I use Interfaces, but I know
you want to make the library easy to port to other languages. So, I'd like
to know if you want to avoid using interfaces, or if that language feature
is ok.
Steve

On Sun, Nov 19, 2017 at 5:12 AM, Wolf Garbe <notifications@github.com>
wrote:

> Thanks for the great refactoring from static to instantiated class!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wolfgarbe/symspell/pull/9#issuecomment-345511941>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACrZBEeNyHM9j9BimwkaynyJUUFr1oQSks5s4BtKgaJpZM4QjPQQ>
> .
>

softwx commented:
Wolf,
Also, while adding a new unit test, I think I may have found a subtle bug
that was present in version 5.0, and still exists. I would have thought
that the following code:

            var symSpell = new SymSpell(1, 3);
            symSpell.CreateDictionaryEntry("pipe", 5);
            symSpell.CreateDictionaryEntry("pips", 10);
            var result = symSpell.Lookup("pipe", 1, 2);

would give a result with two matches (finding both pipe and pips since
verbose is 2, maxEditDistance is 1, and prefixLength is 3. But it only
returns pipe. It skips considering pips for a Damerau comparison because of
this line:

                        //number of edits in prefix ==maxediddistance  AND
no identic suffix, then editdistance>maxEditDistance and no need for
Levenshtein calculation
                        //
(input.Length >= lp) && (suggestion.Length >= lp)
                        if ((prefixLength - maxEditDistance ==
candidate.Length)
                            && (((min = Math.Min(input.Length,
suggestion.Length) - prefixLength) > 1)
                                 && (input.Substring(input.Length + 1 -
min) != suggestion.Substring(suggestion.Length + 1 - min)))
                                || ((min > 0) && (input[input.Length - min]
!= suggestion[suggestion.Length - min])
                                    && ((input[input.Length - min - 1] !=
suggestion[suggestion.Length - min])
                                        || (input[input.Length - min] !=
suggestion[suggestion.Length - min - 1]))))

Am I misinterpreting what those inputs should have produced?
Steve

On Tue, Nov 21, 2017 at 10:03 AM, Steve Hatchett <softwxgit@gmail.com>
wrote:

> Wolf,
> I've worked out a design for a dictionary structure that allows using
> string hashcodes to directly index into the itemList (getting rid of the
> need for the dictionary collection), yet still allow for dynamically adding
> words and growing the structure with a very low cost (fast and simple) size
> doubling  + re-hashing when needed. I think it will be both faster and more
> space efficient, but won't know how well and how much until I implement it.
> The design would also allow some control of the speed vs. space preference.
> It may be easier and cleaner to implement if I use Interfaces, but I know
> you want to make the library easy to port to other languages. So, I'd like
> to know if you want to avoid using interfaces, or if that language feature
> is ok.
> Steve
>
> On Sun, Nov 19, 2017 at 5:12 AM, Wolf Garbe <notifications@github.com>
> wrote:
>
>> Thanks for the great refactoring from static to instantiated class!
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/wolfgarbe/symspell/pull/9#issuecomment-345511941>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ACrZBEeNyHM9j9BimwkaynyJUUFr1oQSks5s4BtKgaJpZM4QjPQQ>
>> .
>>
>
>

wolfgarbe commented:
I don't know how much the new dictionary structure would benefit from interfaces. Both Javascript and Python don't have interfaces, but they are the most actively used ports of SymSpell. And even in the other languages I guess not all developers are familiar with that concept. This will probably lead to fewer or more deviating ports. But I'm not fundamentally opposed.

wolfgarbe commented:
1. Yes, this is probably a bug in that line. I will look into it.

2. Today I tried (among other changes) to make the naming of files and folders more consistent. This included changes of upper/lower case. Probably I messed something up, because suddenly in the github repository appeared two pairs of folders (with different files), one with the old (lower case) and one with the new naming (upper case). Pushing from my consistent local repository did not help. I had to delete the files manually until the double folders disappeared and re-upload those files via github web interface. Unfortunately some of the change history got lost in that transition. Sorry for that.

3. Other changes: 
- Benchmark removed from demo project as there is now a dedicated benchmark project.
- Bug fixed in benchmark: SymSpell() instantiation now right before LoadDirectory(). maxLength was  shared between different dictionaries, this led to wrong result numbers.
- SymSpell.defaultVerbose removed to force people to use and think about verbose, and to tie this parameter to Lookup where it is needed (verbose is only needed in Lookup, not in precalculation).
- Add check to Lookup() whether verbose is in range 0..2
- In lookup() the order of verbose and maxEditDistance parameters has been switched,. Verbose has been made obligatory and maxEditDistance optionally.

4. There is a problem with one of the test cases, at least at my machine. The test "ShouldReplicateNoisyResults" did fail:
Message: System.IO.FileNotFoundException : Could not load file or assembly 'System.IO.FileSystem, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
There might be a version conflict between the net standard 1.3 symspell project and the net framework 4.6.1 test project: https://github.com/nuget/home/issues/4467 
Perhaps changing the test project also to net standard 1.3 could help?

softwx commented:
Wolf,
The line that has the bug is an optimization that won't work with the idea
of moving to hashcodes, so in my current working version, the bug doesn't
exist, as that line goes away. That's how I found it...comparing the old
version results to the new version I'm making, to see if I broke something.

I implemented a first pass at a version that does direct indexing.
Basically it re-implements a custom version of the Dictionary class that
removes some stuff in Dictionary that symspell doesn't need, that just
wastes space, and uses one of the internal structures used in Dictionary to
do what itemList does in the original version of symspell. So, overall,
there is essentially one less large array in memory than before. This was a
quick, naïve implementation just to ensure the idea works (it does). Now
I'll do the optimisations that should make it faster and much more memory
efficient. Initialization with dictionary files should also be much
quicker. I should have it all done in a couple days, as long as I can find
the time. I won't need interfaces after all. The custom collection
structure I'm implementing should be easy port to most languages.

Your verbose change makes sense.

The test failure is because the test isn't finding the dictionary text
file. Moving things around probably changed the relative path. I don't
think it has to do with the .Net version it's built against.

I will merge into my fork all the changes you've made so I don't
accidentally undo any of the work you've been doing. There is a change I
want to make to the benchmark class. On my machine, some of the tests run
fast enough that the times show as .000, so I'd like to bump then to more
digits of precision.

Steve

On Wed, Nov 22, 2017 at 10:21 AM, Wolf Garbe <notifications@github.com>
wrote:

>
>    1.
>
>    Yes, this is probably a bug in that line. I will look into it.
>    2.
>
>    Today I tried (among other changes) to make the naming of files and
>    folders more consistent. This included changes of upper/lower case.
>    Probably I messed something up, because suddenly in the github repository
>    appeared two pairs of folders (with different files), one with the old
>    (lower case) and one with the new naming (upper case). Pushing from my
>    consistent local repository did not help. I had to delete the files
>    manually until the double folders disappeared and re-upload those files via
>    github web interface. Unfortunately some of the change history got lost in
>    that transition. Sorry for that.
>    3.
>
>    Other changes:
>
>
>    - Benchmark removed from demo project as there is now a dedicated
>    benchmark project.
>    - Bug fixed in benchmark: SymSpell() instantiation now right before
>    LoadDirectory(). maxLength was shared between different dictionaries, this
>    led to wrong result numbers.
>    - SymSpell.defaultVerbose removed to force people to use and think
>    about verbose, and to tie this parameter to Lookup where it is needed
>    (verbose is only needed in Lookup, not in precalculation).
>    - Add check to Lookup() whether verbose is in range 0..2
>    - In lookup() the order of verbose and maxEditDistance parameters has
>    been switched,. Verbose has been made obligatory and maxEditDistance
>    optionally.
>
>
>    1. There is a problem with one of the test cases, at least at my
>    machine. The test "ShouldReplicateNoisyResults" did fail:
>    Message: System.IO.FileNotFoundException : Could not load file or
>    assembly 'System.IO.FileSystem, Version=4.0.1.0, Culture=neutral,
>    PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The
>    system cannot find the file specified.
>    There might be a version conflict between the net standard 1.3
>    symspell project and the net framework 4.6.1 test project:
>    https://github.com/nuget/home/issues/4467
>    <https://github.com/nuget/home/issues/4467>
>    Perhaps changing the test project also to net standard 1.3 could help?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wolfgarbe/SymSpell/pull/9#issuecomment-346418632>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACrZBLbt-gxDllW_j46OtrYWW_zr5ii9ks5s5FgugaJpZM4QjPQQ>
> .
>

wolfgarbe commented:
> The test failure is because the test isn't finding the dictionary text file. 
That's what I thought, but after spending quite some time to check path/file locations I disabled/commented the line symSpell.LoadDictionary(path, 0, 1); in the test project and copied its code from the symspell project to the test project:
            using (StreamReader sr = new StreamReader(File.OpenRead(path)))
            {
                String line;

                //process a single line at a time only for memory efficiency
                while ((line = sr.ReadLine()) != null)
                {
                    string[] lineParts = line.Split(null);
                    if (lineParts.Length >= 2)
                    {
                        string key = lineParts[0];
                        //Int64 count;
                        if (Int64.TryParse(lineParts[1], out Int64 count))
                        {
                            symSpell.CreateDictionaryEntry(key, count);
                        }
                    }
                }
            }
And suddenly the test passed (without any change to the path or the location of the file).

Of course we can increase the precision of the benchmark.

wolfgarbe commented:
I investigated and fixed the incomplete result bug. The returned suggestions were not always complete, if maxEditDistance=1 AND input.Length>prefixLength.
Both the input term and the deletes of the prefix of the input term were added as candidate, but not the prefix of the input term itself.
That fooled the early termination, which assumes that candidates are processed in the order of decreasing edit distance.
The bug can also be prevented by disabling the early termination line you mentioned, but then the number of necesary Levenshtein calculations would increase up to factor 4, depending on prefix length (high reduction for small prefix) and dictionary size (high reduction for big dictionaries).
In my tests I'm counting how often the levensthein calculation (as the most expensive part of the lookup) has been called in order to track early termination and other optimzation more precise than with measuring time alone.
I have fixed the bug both in the static SymSpell.Original and the instantiated SymSpell class.

softwx commented:
Wolf,
In the CreateDictionaryEntry method there is a variable called
"countThreshold" whose purpose is to allow a count threshold that a word
must reach to be considered a correct spelling suggestion. I wanted to
ensure that changes I make don't preclude that feature. But while seeing
how that works, I discovered that if you set it to a number greater than 1,
and add a word whose count is below that, the word will not have its
deletes generated (as you would expect), but if you subsequently to a
Lookup of that word, it will still be returned as an exact match. This
doesn't happen when countThreshold is set to 1, because the only count
below 1 is 0, and that seems to be treated internally as a delete word. It
gets more complicated when the threshold is above 1. You might even have a
case where a word is added from the corpus with a low count, and that word
is also a delete of another word.

I have a few questions. 1) is the threshold feature something you want to
keep (it seems you would, since the algorithm puts a fair amount of effort
into maintaining the counts), 2) if yes, should that setting be made an
instance variable that is set in the constructor (but with a default value
of 1), and 3) is it correct or incorrect behavior for a word whose count is
under the threshold to be returned as an exact match?

This is the code I used to test this (plus, the countThreshold has to be
edited in code to be a value >1):
            symSpell = new Original.SymSpell();
            symSpell.CreateDictionaryEntry("pawn", "", 1);
            result = symSpell.Lookup("pawn", "", 0, 0);
            result2 = symSpell.Lookup("pawn", "", 0, 1);

Steve

On Thu, Nov 23, 2017 at 1:15 PM, Wolf Garbe <notifications@github.com>
wrote:

> I investigated and fixed the incomplete result bug. The returned
> suggestions were not always complete, if maxEditDistance=1 AND
> input.Length>prefixLength.
> Both the input term and the deletes of the prefix of the input term were
> added as candidate, but not the prefix of the input term itself.
> That fooled the early termination, which assumes that candidates are
> processed in the order of decreasing edit distance.
> The bug can also be prevented by disabling the early termination line you
> mentioned, but then the number of necesary Levenshtein calculations would
> increase up to factor 4, depending on prefix length (high reduction for
> small prefix) and dictionary size (high reduction for big dictionaries).
> In my tests I'm counting how often the levensthein calculation (as the
> most expensive part of the lookup) has been called in order to track early
> termination and other optimzation more precise than with measuring time
> alone.
> I have fixed the bug both in the static SymSpell.Original and the
> instantiated SymSpell class.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wolfgarbe/SymSpell/pull/9#issuecomment-346692088>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACrZBOHBTFgGbeTy7G8pjwK-05ZOnItOks5s5dJRgaJpZM4QjPQQ>
> .
>

wolfgarbe commented:
Steve,
> 1) is the threshold feature something you want to
> keep (it seems you would, since the algorithm puts a fair amount of effort
> into maintaining the counts), 

When creating a dictionary from a text corpus, that corpus often contains spelling errors. That is especially true for a web corpus, obtained by crawling of random web pages. Another case is dynamically creating or updating a dictionary from user input. The countThreshold prevents adding misspelled words to the dictionary. Of course a simple countThreshold is not perfect, because genuine rare words might be less frequent than common spelling errors of frequent words. I would like to keep the feature.

> 2) if yes, should that setting be made an instance variable 
> that is set in the constructor (but with a default value of 1)

That would be perfect.

> 3) is it correct or incorrect behavior for a word whose count is
> under the threshold to be returned as an exact match?

That is incorrect behaviour. 

Wolf

softwx commented:
I'm attaching a version of SymSpell I made to make sure that my hashcode
idea wasn't just capturing some memory and speed optimisations that could
have been achieved by staying with string values. It does a little
optimization of the current version by combining dictionary and itemList,
but otherwise, it stays very close to your version in how dictionary
entries are created, and how lookups are done. It also fixes the
countThreshold bug I mentioned in my previous email and adds it as a
constructor argument. It's slightly faster in dictionary creation, and a
tiny bit faster in lookups (using Visual Studio's instrumented profiler).
On average it takes less memory to store the dictionary. Sometimes it saves
quite a bit of memory, rarely it consumes a tiny bit more...it depends on
the nature of the words in the dictionary and the construction parameters.
I made it mainly as a base to compare the hashcode implementation against.
After looking at it, if you want me to submit it as a pull request, let me
know.
Steve

On Tue, Nov 28, 2017 at 5:57 AM, Wolf Garbe <notifications@github.com>
wrote:

> Steve,
>
>
>    1. is the threshold feature something you want to
>    keep (it seems you would, since the algorithm puts a fair amount of
>    effort
>    into maintaining the counts),
>
> When creating a dictionary from a text corpus, that corpus often contains
> spelling errors. That is especially true for a web corpus, obtained by
> crawling of random web pages. Another case is dynamically creating or
> updating a dictionary from user input. The countThreshold prevents adding
> misspelled words to the dictionary. Of course a simple countThreshold is
> not perfect, because genuine rare words might be less frequent than common
> spelling errors of frequent words. I would like to keep the feature.
>
>
>    1. if yes, should that setting be made an instance variable
>    that is set in the constructor (but with a default value of 1)
>
> That would be perfect.
>
>
>    1. is it correct or incorrect behavior for a word whose count is
>    under the threshold to be returned as an exact match?
>
> That is incorrect behaviour.
>
> Wolf
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wolfgarbe/SymSpell/pull/9#issuecomment-347515448>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACrZBKZGHDDeL2ZsogP63eaYxDyAoX7wks5s7AMdgaJpZM4QjPQQ>
> .
>

﻿// SymSpell: 1 million times faster through Symmetric Delete spelling correction algorithm
//
// The Symmetric Delete spelling correction algorithm reduces the complexity of edit candidate generation and dictionary lookup 
// for a given Damerau-Levenshtein distance. It is six orders of magnitude faster and language independent.
// Opposite to other algorithms only deletes are required, no transposes + replaces + inserts.
// Transposes + replaces + inserts of the input term are transformed into deletes of the dictionary term.
// Replaces and inserts are expensive and language dependent: e.g. Chinese has 70,000 Unicode Han characters!
//
// Copyright (C) 2017 Wolf Garbe
// Version: 5.1
// Author: Wolf Garbe <wolf.garbe@faroo.com>
// Maintainer: Wolf Garbe <wolf.garbe@faroo.com>
// URL: https://github.com/wolfgarbe/symspell
// Description: http://blog.faroo.com/2012/06/07/improved-edit-distance-based-spelling-correction/
//
// License:
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License, 
// version 3.0 (LGPL-3.0) as published by the Free Software Foundation.
// http://www.opensource.org/licenses/LGPL-3.0

using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

public class SymSpell
{
    public enum Verbosity
    {
        Top,        //the suggestion with the highest term frequency of the suggestions of smallest edit distance found
        Closest,    //all suggestions of smallest edit distance found, the suggestions are ordered by term frequency
        All         //all suggestions <= maxEditDistance, the suggestions are ordered by edit distance, then by term frequency (slower, no early termination)
    };

    const int defaultMaxEditDistance = 2;
    const int defaultPrefixLength = 7;
    const int defaultCountThreshold = 1;

    private int maxDictionaryEditDistance;
    private int maxLength; //maximum dictionary term length
    private int prefixLength; //prefix length  5..7
    private Int64 countThreshold; //a treshold might be specifid, when a term occurs so frequently in the corpus that it is considered a valid word for spelling correction

    public Dictionary<string, DictionaryItem> dictionary = new Dictionary<string, DictionaryItem>(); //initialisierung
    //List of unique words. By using the suggestions (Int) as index for this list they are translated into the original string.
    private List<string> wordlist = new List<string>();

    // if item has no suggestions, Suggestion = -1, and Suggestions = null
    // if item has one suggestion, Suggestion = wordList index, and Suggestions = null
    // if item has more than one suggestion, Suggestion = wordList index of first suggestion, and Suggestions = list of all suggestions (including first)
    public struct DictionaryItem
    {
        public static DictionaryItem Empty = new DictionaryItem(-1, null, -1);
        public int Suggestion { get; private set; } 
        public List<int> Suggestions { get; private set; }
        public Int64 Count { get; private set; }
        public DictionaryItem(int suggestion, List<int> suggestions, Int64 count)
        {
            Suggestion = suggestion;
            Suggestions = suggestions;
            Count = count;
        }
        public override string ToString()
        {
            return "{" + Suggestion + ", " + Suggestions + ", " + Count + "}";
        }
    }

    public class SuggestItem
    {
        public string term = "";
        public int distance = 0;
        public Int64 count = 0;

        public override bool Equals(object obj)
        {
            return Equals(term, ((SuggestItem)obj).term);
        }

        public override int GetHashCode()
        {
            return term.GetHashCode();
        }
        public override string ToString()
        {
            return "{" + term + ", " + distance + ", " + count + "}";
        }
    }

    /// <summary>Maximum edit distance for dictionary precalculation.</summary>
    public int MaxDictionaryEditDistance { get { return this.maxDictionaryEditDistance; } }

    /// <summary>Length of prefix, from which deletes are generated.</summary>
    public int PrefixLength { get { return this.prefixLength; } }

    /// <summary>Length of longest word in the dictionary.</summary>
    public int MaxLength { get { return this.maxLength; } }

    /// <summary>Count threshold for a word to be considered a valid word for spelling correction.</summary>
    public long CountThreshold { get { return this.countThreshold; } }
    
    /// <summary>Number of unique words in the dictionary.</summary>
    public int WordCount { get { return this.wordlist.Count; } }

    /// <summary>Number of words and intermediate word deletes encoded in the dictionary.</summary>
    public int EntryCount { get { return this.dictionary.Count; } }

    public SymSpell(int maxDictionaryEditDistance = defaultMaxEditDistance, int prefixLength = defaultPrefixLength
        , int countThreshold = defaultCountThreshold) {
        this.maxDictionaryEditDistance = maxDictionaryEditDistance;
        this.prefixLength = prefixLength;
        this.countThreshold = countThreshold;
    }

    //for every word there all deletes with an edit distance of 1..maxEditDistance created and added to the dictionary
    //every delete entry has a suggestions list, which points to the original term(s) it was created from
    //The dictionary may be dynamically updated (word frequency and new words) at any time by calling createDictionaryEntry
    public bool CreateDictionaryEntry(string key, Int64 count)
    {
        if (count == 0 && this.countThreshold > 0) return false; // no point doing anything if count is zero, as it can't change anything

        bool result = false;
        Int64 newCount, countPrevious;
        countPrevious = -1;
        DictionaryItem dictItem;
        if (dictionary.TryGetValue(key, out dictItem))
        {
            //summarizes multiple frequency entries of a word (prevents overflow)
            countPrevious = dictItem.Count;
            newCount = (Int64.MaxValue - countPrevious > count) ? countPrevious + count : Int64.MaxValue;

            dictItem = new DictionaryItem(dictItem.Suggestion, dictItem.Suggestions, newCount);
            this.dictionary[key] = dictItem;
        }
        else
        {
            newCount = count;
            dictItem = new DictionaryItem(-1, null, newCount);
            if (key.Length > maxLength) maxLength = key.Length;
        }

        //edits/suggestions are created only once, no matter how often word occurs
        //edits/suggestions are created only as soon as the word occurs in the corpus, 
        //even if the same term existed before in the dictionary as an edit from another word
        if ((newCount >= this.countThreshold) && (countPrevious < this.countThreshold))
        {
            //word2index
            wordlist.Add(key);
            Int32 keyint = (Int32)(wordlist.Count - 1);
            if (dictItem.Suggestion < 0)
            {
                dictItem = new DictionaryItem(keyint, dictItem.Suggestions, dictItem.Count);
                this.dictionary[key] = dictItem;
            }
            else
            {
                if (dictItem.Suggestions == null)
                {
                    dictItem = new DictionaryItem(dictItem.Suggestion, new List<int>(1), dictItem.Count);
                    dictItem.Suggestions.Add(dictItem.Suggestion);
                    this.dictionary[key] = dictItem;
                }
                dictItem.Suggestions.Add(keyint);
            }
            result = true;

            //create deletes
            foreach (string delete in EditsPrefix(key))
            {
                //Int32 value2;
                DictionaryItem di;
                if (dictionary.TryGetValue(delete, out di))
                {
                    if (di.Suggestions == null)
                    {
                        di = new DictionaryItem(di.Suggestion, new List<int>(1), di.Count);
                        di.Suggestions.Add(di.Suggestion);
                        this.dictionary[delete] = di;
                    }
                    di.Suggestions.Add(keyint);
                }
                else
                {
                    dictionary.Add(delete, new DictionaryItem(keyint, null, 0));
                }

            }
        }
        return result;
    }

    //load a frequency dictionary (merges with any dictionary data already loaded)
    public bool LoadDictionary(string corpus, int termIndex, int countIndex)
    {
        if (!File.Exists(corpus)) return false;
        
        using (StreamReader sr = new StreamReader(File.OpenRead(corpus)))
        {
            String line;

            //process a single line at a time only for memory efficiency
            while ((line = sr.ReadLine()) != null)
            {              
                string[] lineParts = line.Split(null);
                if (lineParts.Length>=2)
                {           
                    string key = lineParts[termIndex];
                    //Int64 count;
                    if (Int64.TryParse(lineParts[countIndex], out Int64 count))
                    {
                        CreateDictionaryEntry(key, count);
                    }
                }
            }
        }
        return true;
    }

    //create a frequency dictionary from a corpus (merges with any dictionary data already loaded) 
    public bool CreateDictionary(string corpus)
    {
        if (!File.Exists(corpus)) return false;
        
        using (StreamReader sr = new StreamReader(File.OpenRead(corpus)))
        {
            String line;
            //process a single line at a time only for memory efficiency
            while ((line = sr.ReadLine()) != null)
            {
                foreach (string key in ParseWords(line))
                {
                    CreateDictionaryEntry(key, 1);
                }
            }
        }

        return true;
    }

    public List<SuggestItem> Lookup(string input, Verbosity verbosity)
    {
        return Lookup(input, verbosity, this.maxDictionaryEditDistance);
    }

    public List<SuggestItem> Lookup(string input, Verbosity verbosity, int maxEditDistance)
    {
        //verbosity=Top: the suggestion with the highest term frequency of the suggestions of smallest edit distance found
        //verbosity=Closest: all suggestions of smallest edit distance found, the suggestions are ordered by term frequency 
        //verbosity=All: all suggestions <= maxEditDistance, the suggestions are ordered by edit distance, then by term frequency (slower, no early termination)

        // maxEditDistance used in Lookup can't be bigger than the maxDictionaryEditDistance
        // used to construct the underlying dictionary structure.
        if (maxEditDistance > MaxDictionaryEditDistance) throw new ArgumentOutOfRangeException(nameof(maxEditDistance));

        //save some time
        if (input.Length - maxEditDistance > maxLength) return new List<SuggestItem>();

        List<string> candidates = new List<string>();
        HashSet<string> hashset1 = new HashSet<string>();

        List<SuggestItem> suggestions = new List<SuggestItem>();
        HashSet<string> hashset2 = new HashSet<string>();

        int maxEditDistance2 = maxEditDistance;

        int candidatePointer = 0;

        //add original term
        candidates.Add(input);

        //add original prefix
        if (input.Length > prefixLength) candidates.Add(input.Substring(0, prefixLength));

        while (candidatePointer < candidates.Count)
        {
            string candidate = candidates[candidatePointer++];
            int lengthDiff = Math.Min(input.Length, prefixLength) - candidate.Length;

            //save some time
            //early termination
            //suggestion distance=candidate.distance... candidate.distance+maxEditDistance                
            //if canddate distance is already higher than suggestion distance, than there are no better suggestions to be expected
            if ((verbosity < Verbosity.All) && (suggestions.Count > 0) && (lengthDiff > suggestions[0].distance)) goto sort;

            //read candidate entry from dictionary
            if (dictionary.TryGetValue(candidate, out DictionaryItem dictItem) && dictItem.Suggestion >= 0)
            {
                var dictItemSuggestions = dictItem.Suggestions;
                if (dictItemSuggestions == null)
                {
                    dictItemSuggestions = new List<int>(1);
                    dictItemSuggestions.Add(dictItem.Suggestion);
                }

                //if count>0 then candidate entry is correct dictionary term, not only delete item
                if (dictItem.Count > 0)
                {
                    int distance = input.Length - candidate.Length;

                    //save some time
                    //do not process higher distances than those already found, if verbosity<2      
                    if ((distance <= maxEditDistance)
                    && ((verbosity == Verbosity.All) || (suggestions.Count == 0) || (distance <= suggestions[0].distance))
                    && (hashset2.Add(candidate)))
                    {
                        //Fix: previously not allways all suggestons within maxEditDistance (verbosity=1) or the best suggestion (verbosity=0) were returned : e.g. elove did not return love
                        //suggestions.Clear() was not executed in this branch, if a suggestion with lower edit distance was added here (for verbosity<2). 
                        //Then possibly suggestions with higher edit distance remained on top, the suggestion with lower edit distance were added to the end. 
                        //All of them where deleted later once a suggestion with a lower distance than the first item in the list was later added in the other branch. 
                        //Therefore returned suggestions were not always complete for verbosity<2.
                        //remove all existing suggestions of higher distance, if verbosity<2
                        if ((verbosity < Verbosity.All) && (suggestions.Count > 0) && (suggestions[0].distance > distance)) suggestions.Clear();

                        //add correct dictionary term term to suggestion list
                        SuggestItem si = new SuggestItem()
                        {
                            term = candidate,
                            count = dictItem.Count,
                            distance = distance
                        };
                        suggestions.Add(si);
                        //early termination
                        if ((verbosity < Verbosity.All) && (distance == 0)) goto sort;
                    }
                }

                //iterate through suggestions (to other correct dictionary items) of delete item and add them to suggestion list
                foreach (int suggestionint in dictItemSuggestions)
                {
                    //save some time 
                    //skipping double items early: different deletes of the input term can lead to the same suggestion
                    //index2word
                    string suggestion = this.wordlist[suggestionint];

                    //True Damerau-Levenshtein Edit Distance: adjust distance, if both distances>0
                    //We allow simultaneous edits (deletes) of maxEditDistance on on both the dictionary and the input term. 
                    //For replaces and adjacent transposes the resulting edit distance stays <= maxEditDistance.
                    //For inserts and deletes the resulting edit distance might exceed maxEditDistance.
                    //To prevent suggestions of a higher edit distance, we need to calculate the resulting edit distance, if there are simultaneous edits on both sides.
                    //Example: (bank==bnak and bank==bink, but bank!=kanb and bank!=xban and bank!=baxn for maxEditDistance=1)
                    //Two deletes on each side of a pair makes them all equal, but the first two pairs have edit distance=1, the others edit distance=2.
                    int distance = 0;// maxEditDistance+1;
                    if (suggestion != input)
                    {
                        int min = 0;
                        if (Math.Abs(suggestion.Length - input.Length) > maxEditDistance2)
                        {
                            continue;
                        }
                        else if (candidate.Length == 0)
                        {
                            //suggestions which have no common chars with input (input.length<=maxEditDistance && suggestion.length<=maxEditDistance)
                            if (!hashset2.Add(suggestion)) continue; distance = Math.Max(input.Length, suggestion.Length);
                        }
                        else
                        //number of edits in prefix ==maxediddistance  AND no identic suffix, then editdistance>maxEditDistance and no need for Levenshtein calculation  
                        //                                                 (input.Length >= prefixLength) && (suggestion.Length >= prefixLength) 
                        if ((prefixLength - maxEditDistance == candidate.Length) && (((min = Math.Min(input.Length, suggestion.Length) - prefixLength) > 1) && (input.Substring(input.Length + 1 - min) != suggestion.Substring(suggestion.Length + 1 - min))) || ((min > 0) && (input[input.Length - min] != suggestion[suggestion.Length - min]) && ((input[input.Length - min - 1] != suggestion[suggestion.Length - min]) || (input[input.Length - min] != suggestion[suggestion.Length - min - 1]))))
                        {
                            continue;
                        }
                        else
                        //edit distance of remaining string (after prefix)
                        if ((suggestion.Length == candidate.Length) && (input.Length <= prefixLength)) { if (!hashset2.Add(suggestion)) continue; distance = input.Length - candidate.Length; }
                        else if ((input.Length == candidate.Length) && (suggestion.Length <= prefixLength)) { if (!hashset2.Add(suggestion)) continue; distance = suggestion.Length - candidate.Length; }
                        else if (hashset2.Add(suggestion))
                        {
                            distance = EditDistance.DamerauLevenshteinDistance(input, suggestion, maxEditDistance2); if (distance < 0) distance = maxEditDistance + 1;
                        }
                        else
                        {
                            continue;
                        }
                    }
                    else if (!hashset2.Add(suggestion)) continue;

                    //save some time
                    //do not process higher distances than those already found, if verbosity<2
                    if ((verbosity < Verbosity.All) && (suggestions.Count > 0) && (distance > suggestions[0].distance)) continue;
                    if (distance <= maxEditDistance)
                    {
                        if (dictionary.TryGetValue(suggestion, out DictionaryItem dictItem2))
                        {
                            SuggestItem si = new SuggestItem()
                            {
                                term = suggestion,
                                count = dictItem2.Count,
                                distance = distance
                            };

                            //we will calculate DamLev distance only to the smallest found distance sof far
                            if (verbosity < Verbosity.All) maxEditDistance2 = distance;

                            //remove all existing suggestions of higher distance, if verbosity<2
                            if ((verbosity < Verbosity.All) && (suggestions.Count > 0) && (suggestions[0].distance > distance)) suggestions.Clear();
                            suggestions.Add(si);
                        }
                    }

                }//end foreach
            }//end if         

            //add edits 
            //derive edits (deletes) from candidate (input) and add them to candidates list
            //this is a recursive process until the maximum edit distance has been reached
            if ((lengthDiff < maxEditDistance) && (candidate.Length <= prefixLength))
            {
                //save some time
                //do not create edits with edit distance smaller than suggestions already found
                if ((verbosity < Verbosity.All) && (suggestions.Count > 0) && (lengthDiff >= suggestions[0].distance)) continue;

                for (int i = 0; i < candidate.Length; i++)
                {
                    string delete = candidate.Remove(i, 1);

                    if (hashset1.Add(delete)) { candidates.Add(delete); }
                }
            }
        }//end while

        //sort by ascending edit distance, then by descending word frequency
        sort: if (verbosity < Verbosity.All) suggestions.Sort((x, y) => -x.count.CompareTo(y.count)); else suggestions.Sort((x, y) => 2 * x.distance.CompareTo(y.distance) - x.count.CompareTo(y.count));
        if ((verbosity == 0) && (suggestions.Count > 1)) return suggestions.GetRange(0, 1); else return suggestions;
    }//end if         

    //create a non-unique wordlist from sample text
    //language independent (e.g. works with Chinese characters)
    private string[] ParseWords(string text)
    {
        // \w Alphanumeric characters (including non-latin characters, umlaut characters and digits) plus "_" 
        // \d Digits
        // Compatible with non-latin characters, does not split words at apostrophes
        MatchCollection mc = Regex.Matches(text.ToLower(), @"['’\w-[_]]+");

        //for benchmarking only: with CreateDictionary("big.txt","") and the text corpus from http://norvig.com/big.txt  the Regex below provides the exact same number of dictionary items as Norvigs regex "[a-z]+" (which splits words at apostrophes & incompatible with non-latin characters)     
        //MatchCollection mc = Regex.Matches(text.ToLower(), @"[\w-[\d_]]+");

        var matches = new string[mc.Count];
        for (int i = 0; i < matches.Length; i++) matches[i] = mc[i].ToString();
        return matches;
    }

    //inexpensive and language independent: only deletes, no transposes + replaces + inserts
    //replaces and inserts are expensive and language dependent (Chinese has 70,000 Unicode Han characters)
    private HashSet<string> Edits(string word, int editDistance, HashSet<string> deletes)
    {
        editDistance++;
        if (word.Length > 1)
        {
            for (int i = 0; i < word.Length; i++)
            {
                string delete = word.Remove(i, 1);
                if (deletes.Add(delete))
                {
                    //recursion, if maximum edit distance not yet reached
                    if (editDistance < maxDictionaryEditDistance) Edits(delete, editDistance, deletes);
                }
            }
        }
        return deletes;
    }

    private HashSet<string> EditsPrefix(string key)
    {
        HashSet<string> hashSet = new HashSet<string>();
        if (key.Length <= maxDictionaryEditDistance) hashSet.Add("");

        if (key.Length > prefixLength)
        {
            hashSet.Add(key.Substring(0, prefixLength));
            return Edits(key.Substring(0, prefixLength), 0, hashSet);
        }
        else
        {
            return Edits(key, 0, hashSet);
        }
    }
    private int GetStringHash(string s)
    {
        //return (uint)s.GetHashCode();

        int len = s.Length;
        if (this.prefixLength < len) len = this.prefixLength;

        //return (s.Substring(0, len)).GetHashCode();

        uint hash = 2166136261;
        for (var i = 0; i < len; i++)
        {
            unchecked
            {
                hash ^= s[i];
                hash *= 16777619;
            }
        }
        return (int)hash;

        //uint num = 5381;
        //uint num1 = num;
        //int i = 0;
        //if (this.prefixLength < len) i = len - this.prefixLength;
        //while (true)
        //{
        //    if (i == len) break;
        //    num = (num << 5) + num ^ s[i];
        //    i++;
        //    if (i == len) break;
        //    num1 = (num1 << 5) + num1 ^ s[i];
        //    i++;
        //}
        //return num + num1 * 1566083941;

    }
}

wolfgarbe commented:
Yes, please submit it as a pull request.
I really appreciate your effort to debug and optimize the code and make it more consistent and usable.

Wolf
